inc/class-tmsc-sync.php

```diff
@@ -0,0 +1,215 @@
+			}
+			if ( ! empty( $_POST['tmsc-db-user'] ) ) {
+				$user = sanitize_text_field( wp_unslash( $_POST['tmsc-db-user'] ) );
+				update_option( 'tmsc-db-user', $user, false );
+				self::$tms_db_user = $user;
+			}
+			if ( ! empty( $_POST['tmsc-db-password'] ) ) {
+				$password = sanitize_text_field( wp_unslash( $_POST['tmsc-db-password'] ) );
+				update_option( 'tmsc-db-password', $password, false );
+				self::$tms_db_password = $password;
+			}
+			if ( ! empty( $_POST['tmsc-image-url'] ) ) {
+				$url = esc_url_raw( wp_unslash( $_POST['tmsc-image-url'] ) );
+				update_option( 'tmsc-image-url', $url, false );
+				self::$image_url = $url;
+			}
+			/**
+			 * Uncomment the schedule event function and comment the object sync function to enable asynchronous sync.
+			 */
+			self::$instance->object_sync();
+			// wp_schedule_single_event( time(), 'tmsc_cron_events', array() );
+			echo 1;
+		} else {
+			echo 0;
+			wp_die( esc_html__( 'You do not have sufficient permissions to access this page.', 'tmsc' ) );
+		}
+		exit();
+	}
+
+	/**
+	 * A generic ajax responder that spits back a option value or returns a boolean if a comparison value is passed.
+	 */
+	public function ajax_get_option_value() {
+		check_ajax_referer( 'wp_admin_js_script', 'nonce' );
+		if ( ! empty( $_POST['option_key'] ) ) {
+			// If an old value is passed, return success only if it is different from the current value.
+			// This can be used for asynchrounous checks.
+			$current_value = get_option( sanitize_key( $_POST['option_key'] ) );
+			if ( ! empty( $_POST['old_value'] ) ) {
+				if ( $_POST['old_value'] !== $current_value && __( 'Syncing TMS Objects', 'tmsc' ) !== $current_value ) {
+					wp_send_json( array( 'success' => true, 'data' => esc_js( $current_value ) ) );
+				} else {
+					wp_send_json_error( $current_value );
+				}
+			} else {
+				wp_send_json_error( $current_value );
+			}
+		} else {
+			wp_send_json_error();
+		}
+		exit();
+	}
+
+	/**
+	 * Add custom intervals
+	 */
+	public function add_intervals( $schedules ) {
+		$schedules['halfhour'] = array(
+			'interval' => 1800,
+			'display' => __( 'Every Half Hour', 'tmsc' ),
+		);
+		return $schedules;
+	}
+
+	public function cron_events_activation() {
+		// Run our Cron every 30 mins.
+		if ( ! wp_next_scheduled( 'tmsc_cron_events' ) ) {
+			wp_schedule_event( time(), 'halfhour', 'tmsc_cron_events' );
+		}
+	}
+
+	/**
+	 * Add in any cron events that need to be fired every 30 mins here.
+	 * @param array. $args. An array of args to pass to this event scheduler hook.
+	 * @return void.
+	 */
+	public function cron_events( $args = array() ) {
+		// Make sure sync is not currently running
+		if ( 'Syncing TMS Objects' !== get_option( 'tmsc-last-sync-date' ) ) {
+			// Update our custom post type.
+			self::$instance->object_sync();
+		}
+	}
+
+	// Connect to the feed and update our post types with the latest data.
+	public function object_sync() {
+
+		ini_set( 'max_execution_time', 300 );
```

Comment:
This seems a bit hacky and if we open source this in the future it will need to be removed.  Is there any way to break out the cron jobs into smaller chunks to not run into the max execution time limit?